impl/src/metadata.rs

```diff
@@ -59,34 +67,153 @@ impl Default for CargoMetadataFetcher {
-      fs::copy(files.toml_path.as_path(), new_toml_path)?;
-      if let Some(lock_path) = files.lock_path_opt.as_ref() {
-        let new_lock_path = dir_path.join(lock_path.file_name().unwrap());
-        fs::copy(lock_path.as_path(), new_lock_path)?;
-      }
+    let cargo_dir = match use_tempdir {
+      false => files.toml_path.as_path().parent().unwrap(),
+      true => {
+        let dir_path = tempdir.path();
+        let new_toml_path = dir_path.join(files.toml_path.file_name().unwrap());
+        fs::copy(files.toml_path.as_path(), new_toml_path)?;
+        if let Some(lock_path) = files.lock_path_opt.as_ref() {
+          let new_lock_path = dir_path.join(lock_path.file_name().unwrap());
+          fs::copy(lock_path.as_path(), new_lock_path)?;
+        }
 
-      dir
+        dir_path
+      },
     };
 
     MetadataCommand::new()
       .cargo_path(&self.cargo_bin_path)
-      .current_dir(cargo_tempdir.path())
+      .current_dir(cargo_dir)
       .exec()
       .map_err(|e| e.into())
   }
 }
 
+fn clone_crate_src(registry_url: &str, dir: &Path, name: &str, version: &str) -> Result<PathBuf> {
+  let url = url::Url::parse(registry_url)?;
+
+  println!("Cloning binary dependency: {}", &name);
+  let mut cloner = cargo_clone::Cloner::new();
+  cloner
+    .set_registry_url(
+      format!(
+        "{}://{}",
+        url.scheme(),
+        url.host_str().ok_or(anyhow!(
+          "A host string is expected in the settings.registry url"
+        ))?
+      )
+      .as_str(),
+    )
+    .set_out_dir(dir);
+
+  cloner.clone(
+    cargo_clone::CloneMethodKind::Crate,
+    name,
+    Some(version),
+    &Vec::new(),
+  )?;
+
+  let crate_dir = dir.join(format!("{}-{}", name, version));
+  match crate_dir.exists() {
+    false => Err(anyhow!("Directory does not exist")),
+    true => Ok(crate_dir),
+  }
+}
+
+/** Ensures a lockfile is generated for a crate on disk
+ *
+ * If a lockfile for the current crate exists in the `raze` lockfiles output directory, that
+ * lockfile will instead be installed into the crate's directory. If no lockfile exists, one
+ * will be generated.
+ */
+fn generate_lockfile(
+  crate_dir: &Path,
+  lockfile_dir: &Path,
+  name: &str,
+  version: &str,
+) -> Result<PathBuf> {
+  let target_lockfile = lockfile_dir.join(format!("Cargo.{}-{}.lock", name, version));
+  let bundled_lockfile = crate_dir.join("Cargo.lock");
+
+  // If a lockfile exists from a previous run, use that instead
+  match target_lockfile.exists() {
+    true => {
+      fs::copy(&target_lockfile, &bundled_lockfile)?;
+    },
+    false => {
+      // Generate a new lockfile if one wasn't included in the bundle
+      if !bundled_lockfile.exists() {
+        std::process::Command::new(SYSTEM_CARGO_BIN_PATH)
+          .arg("generate-lockfile")
+          .current_dir(&crate_dir)
+          .output()?;
+      }
+    },
+  }
+
+  Ok(bundled_lockfile)
+}
+
+pub fn fetch_crate_src(
```

Comment:
Happy to move it but I'd like to hear more about why you don't feel this belongs in `metadata`. I ask trying to prevent `util` from becoming a dumping ground for random functionality.

Reply:
So, caveat on below. I don't think I've constructed a comprehensive mental model for how modules should be delineated in rust, even given the moderate amount of rust code I've written in my life, so everything below is basically "gut feeling", not real principles.

---

In my "platonic ideal world of abstractions", this module is just concerned with the mechanics of loading and extracting metadata about an ensemble of crates. While I do think I would also decide to put functions used by this module which aren't apparently directly related to into it, this function triggered my "this code is tangential" alarm because it was just concerned with fetching remote crates, and it wasn't actually directly used by this module (except in tests). Instead, IIRC, it's used by `planning`. That's a red flag for me.

FWIW, this leads me to a more substantial comment. My intent with the separation of "metadata fetch" and "planning" was to allow all the yucky side-effects like crate fetching and cargo interactions to happen in one black box, and then have "planning" code which was effectively functional over given state in determining what has to get done. Seeing that you're doing things like fetching crates in plan code, we've clearly drifted from that ideal since I originally structured this.

Reply:
Thanks for taking the time to share 😄 

While the changes are still in `metadata` right now, I think I've updated the code to better align with your concerns here (though likely still not the ideal state). Is this a step in the right direction?